src/lib.rs

```diff
@@ -570,14 +574,82 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'tcx> {
     fn stack_push(
         ecx: &mut InterpretCx<'mir, 'tcx, Self>,
     ) -> InterpResult<'tcx, stacked_borrows::CallId> {
-        Ok(ecx.memory().extra.borrow_mut().new_call())
+        Ok(ecx.memory().extra.stacked.borrow_mut().new_call())
     }
 
     #[inline(always)]
     fn stack_pop(
         ecx: &mut InterpretCx<'mir, 'tcx, Self>,
         extra: stacked_borrows::CallId,
     ) -> InterpResult<'tcx> {
-        Ok(ecx.memory().extra.borrow_mut().end_call(extra))
+        Ok(ecx.memory().extra.stacked.borrow_mut().end_call(extra))
+    }
+
+    fn int_to_ptr(
+        int: u64,
+        memory: &Memory<'mir, 'tcx, Self>,
+    ) -> InterpResult<'tcx, Pointer<Self::PointerTag>> {
+        if int == 0 {
+            return err!(InvalidNullPointerUsage);
+        }
+        
+        if memory.extra.rng.is_none() {
+            return err!(ReadBytesAsPointer);
+        }
+
+        let extra = memory.extra.intptrcast.borrow();
+        
+        match extra.int_to_ptr_map.binary_search_by_key(&int, |(int, _)| *int) {
+            Ok(pos) => {
+                let (_, alloc_id) = extra.int_to_ptr_map[pos];
+                Ok(Pointer::new_with_tag(alloc_id, Size::from_bytes(0), Tag::Untagged))
+            }
+            Err(pos) => {
+                if pos > 0 {
+                    let (glb, alloc_id) = extra.int_to_ptr_map[pos - 1];
+                    let offset = int - glb;
+                    if offset <= memory.get(alloc_id)?.bytes.len() as u64 {
+                        Ok(Pointer::new_with_tag(alloc_id, Size::from_bytes(offset), Tag::Untagged))
+                    } else {
+                        return err!(DanglingPointerDeref);
+                    }
+                } else {
+                    return err!(DanglingPointerDeref);
+                }
+            }
+        }
+    }
+ 
+    fn ptr_to_int(
+        ptr: Pointer<Self::PointerTag>,
+        memory: &Memory<'mir, 'tcx, Self>,
+    ) -> InterpResult<'tcx, u64> {
+        if memory.extra.rng.is_none() {
+            return err!(ReadPointerAsBytes);
+        }
+
+        let mut extra = memory.extra.intptrcast.borrow_mut();
+
+        let alloc = memory.get(ptr.alloc_id)?;
+
+        let base_addr = match alloc.extra.base_addr.get() { 
+            Some(base_addr) => base_addr,
+            None => {
+                let base_addr = extra.next_base_addr;
+                extra.next_base_addr += alloc.bytes.len() as u64;
+
+                alloc.extra.base_addr.set(Some(base_addr));
+
+                let elem = (base_addr, ptr.alloc_id);
+
+                // Given that `next_base_addr` increases in each allocation, pushing the
+                // corresponding tuple keeps `int_to_ptr_map` sorted
+                extra.int_to_ptr_map.push(elem); 
+
+                base_addr
+            }
+        };
+
+        Ok(base_addr + ptr.offset.bytes())
```

Comment:
All of the logic here in int_to_ptr and ptr_to_int should be in `intptrcast.rs`.